File: Codes in C++/Dynamic Programming/Longest Common Subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int m=text1.size();
        int n=text2.size();
        vector<int> prev(n+1,0);
        //Intialise
        for(int i=0;i<=n;i++) prev[i]=0;

        for(int i=1;i<=m;i++)
        {   vector<int> cur(n+1,0);
            for(int j=1;j<=n;j++)
            {
                if(text1[i-1]==text2[j-1]) 
                    cur[j]=1+prev[j-1];
                else
                    cur[j]=max(prev[j],cur[j-1]);
            }
            prev=cur;
        }
        return prev[n];
    }
};



// Print one LCS
        //  int len = dp[n][m];
        //   int i = n;
        //   int j = m;
        
        //   int index = len - 1;
        //   string str = "";
        //   for (int k = 1; k <= len; k++) {
        //     str += "$"; // dummy string
        //   }
        
        //   while (i > 0 && j > 0) {
        //     if (s1[i - 1] == s2[j - 1]) {
        //       str[index] = s1[i - 1];
        //       index--;
        //       i--;
        //       j--;
        //     } else if (s1[i - 1] > s2[j - 1]) {
        //       i--;
        //     } else j--;
        //   }
		// }




// class Solution
// {
// public:
//     int longestCommonSubsequence(string text1, string text2)
//     {
//         int m = text1.length();
//         int n = text2.length();
//         int LCS[m + 1][n + 1];
//         for (int i = 0; i < m + 1; i++)
//         {
//             for (int j = 0; j < n + 1; j++)
//             {
//                 LCS[i][j] = 0;
//             }
//         }
//         for (int i = 1; i <= m; i++)
//             for (int j = 1; j <= n; j++)
//             {
//                 if (text1[i - 1] == text2[j - 1])
//                     LCS[i][j] = LCS[i - 1][j - 1] + 1;
//                 else
//                     LCS[i][j] = max(LCS[i - 1][j], LCS[i][j - 1]);
//             }
//         return LCS[m][n];
//     }
// };

File: Codes in Python3/Dynamic Programming/Longest Common Subsequence.py
#DP Space Optimised Bottom-Up Approach
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        if len(text1)<len(text2):
            text1,text2=text2,text1
        #Whichever is shorter string that size dp array should
        #be created to optimised space
        prev=[0]*(len(text2)+1) 
        
        for i in range(1,len(text1)+1):
            curr=[0]*(len(text2)+1) 
            for j in range(1,len(text2)+1):
                if text1[i-1]==text2[j-1]:
                    curr[j]=prev[j-1]+1
                else:
                    curr[j]=max(curr[j-1],prev[j])  
            prev=curr       

        return prev[len(text2)]
# TC=O(m*n)
# SC=O(min(m,n))
        
#DP Bottom-Up Approach
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp=[[0]*(len(text2)+1) for _ in range(len(text1)+1)]
        
        for i in range(1,len(text1)+1):
            for j in range(1,len(text2)+1):
                if text1[i-1]==text2[j-1]:
                    dp[i][j]=dp[i-1][j-1]+1
                else:
                    dp[i][j]=max(dp[i][j-1],dp[i-1][j])         

        return dp[len(text1)][len(text2)]
# TC=O(m*n)
# SC=O(m*n)
        
#DP Top-Down Approach
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp=[[-1]*(len(text2)+1) for _ in range(len(text1)+1)]
        
        def f(i,j):
            if i not in range(0,len(text1)) or j not in range(0,len(text2)):
                return 0

            if dp[i][j]!=-1:
                return dp[i][j]

            if text1[i]==text2[j]:
                dp[i][j]=f(i-1,j-1)+1
            else:
                dp[i][j]=max(f(i,j-1),f(i-1,j))
            return dp[i][j]

        return f(len(text1)-1,len(text2)-1)
# TC=O(m*n)
# SC=O(m*n)
        

#Recursion
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        
        def f(i,j):
            if i not in range(0,len(text1)) or j not in range(0,len(text2)):
                return 0
            res=-float("inf")
            if text1[i]==text2[j]:
                res=f(i-1,j-1)+1
            else:
                res=max(f(i,j-1),f(i-1,j))
            return res
        ans=f(len(text1)-1,len(text2)-1)
        return ans
# TC=O(2^(m+n))
# SC=O(m+n)
        